opm/grid/cpgrid/CpGridData.hpp

```diff
@@ -203,18 +203,17 @@ class CpGridData
     /// \param turn_normals if true, all normals will be turned. This is intended for handling inputs with wrong orientations.
     /// \param clip_z if true, the grid will be clipped so that the top and bottom will be planar.
     /// \param poreVolume pore volumes for use in MINPV processing, if asked for in deck
-    void processEclipseFormat(const Opm::EclipseGrid* ecl_grid, bool periodic_extension, bool turn_normals = false, bool clip_z = false,
-                              const std::vector<double>& poreVolume = std::vector<double>(), const Opm::NNC& nncs = Opm::NNC(),
-                              const std::unordered_map<size_t, double>& aquifer_cell_volumes = std::unordered_map<size_t, double>());
+    void processEclipseFormat(Opm::EclipseState& ecl_state, const Opm::Deck& deck,
+                              bool periodic_extension, bool turn_normals = false, bool clip_z = false);
 #endif
 
     /// Read the Eclipse grid format ('grdecl').
     /// \param input_data the data in grdecl format, declared in preprocess.h.
     /// \param z_tolerance points along a pillar that are closer together in z
     ///        coordinate than this parameter, will be replaced by a single point.
     /// \param remove_ij_boundary if true, will remove (i, j) boundaries. Used internally.
-    void processEclipseFormat(const grdecl& input_data, const std::array<std::set<std::pair<int, int>>, 2>& nnc, double z_tolerance, bool remove_ij_boundary, bool turn_normals = false,
-                              const std::unordered_map<size_t, double>& aquifer_cell_volumes = std::unordered_map<size_t, double>());
+    void processEclipseFormat(const grdecl& input_data, Opm::EclipseState& ecl_state, const Opm::Deck& deck,
```

Comment:
Your really need to document this properly, as the change seems a bit dangerous. We seem to call this function with an empty ecl_state, which is fine as long as there are no numerical aquifers.

The danger is that in the future fellow developer might assume that the ecl_state is always valid and use it for other things. In that case they will have a hard time debugging.

Reply:
This looks very dangerous to me - the empty `EclipseState` is purely an artifact of the serialization - it is deeply invalid and should not be used for anything!

Reply:
I will add the document here. If we do not use `EclipseState` here, we will need the `EclipseGrid`, `FieldPropsManager`, `AquiferConfig` and `inputNNC` of the `EclipseState`. I really do not want to add so many arguments and passing them around. One of the big difficulties when developing this part is that we have five different functions with the same name `processEclipseFormat`, with different arguments, and doing different but related things, and call one other.  Maybe it is good idea to rename the functions to get a little bit more descriptive (maybe I should have done it before developing this PR). Anyway, it is not related to this PR directly. 

> We seem to call this function with an empty ecl_state,

What do you mean here?

Reply:
The empty `ecl_state` happens in the following code in CpGrid.cpp. 
```c++
    void CpGrid::processEclipseFormat(const grdecl& input_data, double z_tolerance,
                                      bool remove_ij_boundary, bool turn_normals)
    {
        // TODO: just fix the interface, not sure it will be problematic
        Opm::EclipseState ecl_state;
        Opm::Deck deck;
        using NNCMap = std::set<std::pair<int, int>>;
        using NNCMaps = std::array<NNCMap, 2>;
        NNCMaps nnc;
        current_view_data_->processEclipseFormat(input_data, ecl_state, deck, nnc, z_tolerance, remove_ij_boundary, turn_normals);
```
Do you think to use pointers here better than using the references? Then we can just pass in `nullptr` for the above situation.  Please comment.

Reply:
Yes I think using a pointer or optional is better here - as I said - things will go fast to pieces if you use a default constructed EclipseState

Reply:
Okay. I will use pointers here. I think optional to reference is not there yet, right? Correct me if I am wrong here.

Reply:
Wouldn't std::optional make a copy of EclipseState? That might be expensive...
You would need to use a std::optional with a std::reference_wrapper<EclipseState> inside to prevent the copy. Not sure whether that really works.

Reply:
>You would need to use a std::optional with a std::reference_wrapper inside to prevent the copy. Not sure whether that really works.

That works - I have tried it in another context; but maybe slightly over engineered? I would go with a pointer, but this is not my turf.

Reply:
This change really backfired in the downstream modules. in opm-upscaling module, it only looks at the eclipse_grid. With the same grid deck, you can generate a EclipseGrid and also possible for a EclipseState with an EclipseGrid. But these two grid are different, because since there is no pore volumes, all the grid cells will be marked in active.